internal/querycoord/global_meta_broker.go

```diff
@@ -267,41 +269,152 @@ func (broker *globalMetaBroker) parseIndexInfo(ctx context.Context, segmentID Un
-		if err != nil {
-			return nil, err
+	ret := make(map[UniqueID][]*querypb.VecFieldIndexInfo)
+	for _, segmentID := range segmentIDs {
+		infos, ok := resp.GetSegmentInfos()[segmentID]
+		if !ok {
+			log.Warn("segment not found",
+				zap.Int64("collection", collectionID),
+				zap.Int64("segment", segmentID))
+			return nil, fmt.Errorf("segment not found, collection: %d, segment: %d", collectionID, segmentID)
+		}
+
+		if _, ok := ret[segmentID]; !ok {
+			ret[segmentID] = make([]*querypb.VecFieldIndexInfo, 0, len(infos.IndexInfos))
+		}
+
+		for _, info := range infos.IndexInfos {
+			extraInfo, ok := infos.GetExtraIndexInfos()[info.IndexID]
+			indexInfo := &querypb.VecFieldIndexInfo{
+				FieldID:        info.FieldID,
+				EnableIndex:    info.EnableIndex,
+				IndexName:      "",
+				IndexID:        info.IndexID,
+				BuildID:        info.BuildID,
+				IndexParams:    nil,
+				IndexFilePaths: nil,
+				IndexSize:      0,
+			}
+
+			if !info.EnableIndex {
+				ret[segmentID] = append(ret[segmentID], indexInfo)
+				continue
+			}
+
+			paths, err := broker.getIndexFilePaths(ctx, info.BuildID)
+			if err != nil {
+				log.Error("failed to get index file paths",
+					zap.Int64("collection", collectionID),
+					zap.Int64("segment", segmentID),
+					zap.Int64("buildID", info.BuildID),
+					zap.Error(err))
+				return nil, err
+			}
+
+			if len(paths) <= 0 || len(paths[0].IndexFilePaths) <= 0 {
+				log.Warn("index not ready", zap.Int64("index_build_id", info.BuildID))
+				return nil, fmt.Errorf("index not ready, index build id: %d", info.BuildID)
+			}
+
+			indexInfo.IndexFilePaths = paths[0].IndexFilePaths
+			indexInfo.IndexSize = int64(paths[0].SerializedSize)
+
+			if ok {
+				indexInfo.IndexName = extraInfo.IndexName
+				indexInfo.IndexParams = extraInfo.IndexParams
+			} else {
+				// get index name, index params from binlog.
+				extra, err := broker.loadIndexExtraInfo(ctx, paths[0])
+				if err != nil {
+					log.Error("failed to load index extra info",
+						zap.Int64("index build id", info.BuildID),
+						zap.Error(err))
+					return nil, err
+				}
+				indexInfo.IndexName = extra.indexName
+				indexInfo.IndexParams = extra.indexParams
+			}
+
+			ret[segmentID] = append(ret[segmentID], indexInfo)
 		}
 	}
-	log.Debug("get index info success", zap.Int64("collectionID", collectionID), zap.Int64("segmentID", segmentID), zap.Bool("enableIndex", enableIndex))
 
-	return []*querypb.VecFieldIndexInfo{indexInfo}, nil
+	return ret, nil
+}
+
+func (broker *globalMetaBroker) getFullIndexInfo(ctx context.Context, collectionID UniqueID, segmentID UniqueID) ([]*querypb.VecFieldIndexInfo, error) {
+	segmentIndexInfos, err := broker.getFullIndexInfos(ctx, collectionID, []UniqueID{segmentID})
+	if err != nil {
+		return nil, err
+	}
+	if infos, ok := segmentIndexInfos[segmentID]; ok {
+		return infos, nil
+	}
+	return nil, fmt.Errorf("failed to get segment index infos, collection: %d, segment: %d", collectionID, segmentID)
+}
+
```

Comment:
this function does not need to exist, why not use broker.getFullIndexInfos directly?

Reply:
OK